src/backend/lang_c/c_simple_ast_builder.rs

```diff
@@ -0,0 +1,371 @@
+                } else {
+                    unreachable!();
+                }
+            } else if self.ssa.is_block(node) {
+                last = Some(node);
+            }
+        }
+    }
+
+    pub fn recover_code(rfn: &'a RadecoFunction) -> SimpleCAST {
+        let mut builder = Self::new(rfn);
+        // Recover values
+        let data_graph = CASTDataMap::recover_data(rfn, &mut builder.ast);
+        builder.datamap = data_graph;
+        builder.declare_vars();
+        // Recover control flow graph
+        builder.cfg_from_blocks(builder.ssa.entry_node().unwrap(), &mut HashSet::new());
+        builder.replace_tmp_with_goto();
+        builder.ast
+    }
+
+    fn cfg_from_nodes(&mut self, block: NodeIndex) {
+        let nodes = self.ssa.nodes_in(block);
+        for node in nodes {
+            if self.is_recover_action(node) {
+                let n = self.recover_action(node);
+                self.action_map.insert(node, n);
+            }
+        }
+    }
+
+    fn cfg_from_blocks(&mut self, block: NodeIndex, visited: &mut HashSet<NodeIndex>) {
+        if visited.contains(&block) {
+            return;
+        }
+        visited.insert(block);
+        let next_blocks = self.ssa.next_blocks(block);
+        for blk in next_blocks {
+            let n = self.dummy_goto();
+            self.action_map.insert(blk, n);
+            self.cfg_from_nodes(blk);
+            self.cfg_from_blocks(blk, visited);
+        }
+    }
+}
+
+pub struct CASTDataMap<'a> {
+    rfn: &'a RadecoFunction,
+    ssa: &'a SSAStorage,
+    // Hashmap from node of SSAStorage to one of self.data_graph
+    // a map from node of data_graph to one of SimpleCAST's value
+    pub var_map: HashMap<NodeIndex, NodeIndex>,
+    // a map from node of data_graph to one of SimpleCAST's register
+    pub reg_map: HashMap<String, NodeIndex>,
+    pub consts: HashSet<String>,
+    seen: HashSet<NodeIndex>,
+}
+
+impl<'a> CASTDataMap<'a> {
+    pub fn new(rfn: &'a RadecoFunction) -> CASTDataMap<'a> {
+        CASTDataMap {
+            ssa: rfn.ssa(),
+            rfn: rfn,
+            var_map: HashMap::new(),
+            reg_map: HashMap::new(),
+            consts: HashSet::new(),
+            seen: HashSet::new(),
+        }
+    }
+
+    // Returns data map from SSAStorage's NodeIndex to SimpleCAST's NodeIndex
+    pub fn recover_data(rfn: &'a RadecoFunction, ast: &mut SimpleCAST) -> Self {
+        let mut s = Self::new(rfn);
+        s.prepare_consts(ast);
+        s.prepare_regs(ast);
+        for node in s.ssa.inorder_walk() {
+            if s.ssa.is_phi(node) {
+                s.handle_phi(node);
+            } else if s.ssa.is_expr(node) {
+                s.update_values(node, ast);
+            }
+        }
+        s
+    }
+
+    fn handle_binop(&mut self, ret_node: NodeIndex, ops: Vec<NodeIndex>,
+                    expr: c_simple::Expr, ast: &mut SimpleCAST) {
+        // XXX for debbuging, do not use filter_map
```

Comment:
Why not use `filter_map` here?

Reply:
If the length of result of `filter_map` is different from the one of `ops`, we can not recover expression correctly. So None node has to be replaced with `unknown` node.